iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+    // LHS and RHS vectors, `lhsI8` and `rhsI8`.
+    if (!isMatrixTimesMatrixTransposed(contractionOp, 8, 4, 8)) {
+      return failure();
+    }
+    Type I8Type = rewriter.getIntegerType(8);
+    Type I32Type = rewriter.getIntegerType(32);
+    VectorType accType = contractionOp.acc().getType().cast<VectorType>();
+    if (accType.getElementType() != I32Type) {
+      return failure();
+    }
+    Value lhsI8 = getExtSIInput(I8Type, I32Type, contractionOp.lhs());
+    Value rhsI8 = getExtSIInput(I8Type, I32Type, contractionOp.rhs());
+    if (!lhsI8 || !rhsI8) return failure();
+
+    // `contractionOp` matches, start rewriting it. We only reference
+    // the `lhsI8` and `rhsI8` values obtained above as the inputs of the
+    // arith.extsi, so this rewrite will leave the existing arith.extsi without
+    // any user (unless something else was using them), so they may be
+    // removed by another transformation.
+    Location loc = contractionOp.getLoc();
+    // Flatten the inputs to 1D vectors.
+    Value flatLhsI8 = flatten(rewriter, loc, lhsI8);
+    Value flatRhsI8 = flatten(rewriter, loc, rhsI8);
+    Value flatAcc = flatten(rewriter, loc, contractionOp.acc());
+
+    // Create the 1D input vectors of 16 bytes each that are directly what
+    // the target SIMD instructions will want.
+    SmallVector<Value> lhsVec;
+    SmallVector<Value> rhsVec;
+    VectorType vector16xi8Type = VectorType::get({16}, I8Type);
+    for (int position = 0; position < 8 * 4; position += 16) {
+      lhsVec.push_back(
+          extract1DSlice(rewriter, loc, vector16xi8Type, flatLhsI8, position));
+      rhsVec.push_back(
+          extract1DSlice(rewriter, loc, vector16xi8Type, flatRhsI8, position));
+    }
+    SmallVector<Value> accVec;
+    VectorType int32x4Type = VectorType::get({4}, I32Type);
+    for (int position = 0; position < 8 * 8; position += 4) {
+      accVec.push_back(
+          extract1DSlice(rewriter, loc, int32x4Type, flatAcc, position));
+    }
+
+    // Start of the code that's specific to inline assembly. An intrinsics
+    // code path would diverge here.
+
+    // Create the inline asm op's operands list.
+    SmallVector<Value> asmOperands;
+    // First the inputs operands.
+    asmOperands.append(lhsVec);
+    asmOperands.append(rhsVec);
+    // Then the input-output operands.
+    asmOperands.append(accVec);
+    SmallVector<Type> asmOutputOperandTypes(
+        llvm::map_range(accVec, [](Value v) { return v.getType(); }));
+
+    // Create the inline asm op.
+    auto returnType = LLVM::LLVMStructType::getLiteral(rewriter.getContext(),
+                                                       asmOutputOperandTypes);
+    auto dialectAttr = LLVM::AsmDialectAttr::get(rewriter.getContext(),
+                                                 LLVM::AsmDialect::AD_ATT);
+    LLVM::InlineAsmOp asmOp = rewriter.create<LLVM::InlineAsmOp>(
+        loc, returnType, asmOperands,
+        R"ASM(
```

Comment:
The `ASM` here is an arbitrary delimiter for c++ multi-string literal string. It's not included in the emitted literal string, the actual string contents starts after the `(`. Does that address your concern? In this instance I think we can omit the delimiter (because the substring `)"` does not occur anywhere in the literal string itself). Would you prefer that?

Reply:
ok, dropped the `ASM` delimiter (it's not needed for now. the substring `)"` is not a common thing to have in assembly, at least given that we are unlikely to want to have comments in that inline assembly).

Reply:
I actually learned after making this comment that this kind of delimiter can be a hint for tooling, like syntax highlighting within an individual file. And it seems like `ASM(` is a common thing to use as the delimiter, so I would actually recommend keeping it. Sorry :grimacing: